File: jacodb-panda-dynamic/src/main/kotlin/org/jacodb/panda/dynamic/ets/dto/Stmts.kt
/*
 *  Copyright 2022 UnitTestBot contributors (utbot.org)
 * <p>
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 * <p>
 *  http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package org.jacodb.panda.dynamic.ets.dto

import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonClassDiscriminator
import kotlinx.serialization.json.JsonElement

@Serializable
@OptIn(ExperimentalSerializationApi::class)
@JsonClassDiscriminator("_")
sealed interface StmtDto

@Serializable
@SerialName("UNKNOWN")
data class UnknownStmtDto(
    val stmt: JsonElement,
) : StmtDto

@Serializable
@SerialName("NopStmt")
object NopStmtDto : StmtDto {
    override fun toString(): String = javaClass.simpleName
}

@Serializable
@SerialName("AssignStmt")
data class AssignStmtDto(
    val left: ValueDto, // Local
    val right: ValueDto,
) : StmtDto {
    override fun toString(): String {
        return "$left = $right"
    }
}

@Serializable
@SerialName("CallStmt")
data class CallStmtDto(
    val expr: CallExprDto,
) : StmtDto {
    override fun toString(): String {
        return expr.toString()
    }
}

@Serializable
@SerialName("DeleteStmt")
data class DeleteStmtDto(
    val arg: FieldRefDto,
) : StmtDto

@Serializable
sealed interface TerminatingStmtDto : StmtDto

@Serializable
@SerialName("ReturnVoidStmt")
object ReturnVoidStmtDto : TerminatingStmtDto {
    override fun toString(): String = javaClass.simpleName
}

@Serializable
@SerialName("ReturnStmt")
data class ReturnStmtDto(
    val arg: ValueDto,
) : TerminatingStmtDto {
    override fun toString(): String {
        return "return $arg"
    }
}

@Serializable
@SerialName("ThrowStmt")
data class ThrowStmtDto(
    val arg: ValueDto,
) : TerminatingStmtDto

@Serializable
sealed interface BranchingStmtDto : StmtDto

@Serializable
@SerialName("GotoStmt")
object GotoStmtDto : BranchingStmtDto {
    override fun toString(): String = javaClass.simpleName
}

@Serializable
@SerialName("IfStmt")
data class IfStmtDto(
    val condition: ConditionExprDto,
) : BranchingStmtDto

@Serializable
@SerialName("SwitchStmt")
data class SwitchStmtDto(
    val arg: ValueDto,
    val cases: List<ValueDto>,
) : BranchingStmtDto

File: jacodb-ets/ts-frontend/src/dto/stmts.ts
/*
 *  Copyright 2022 UnitTestBot contributors (utbot.org)
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

/**
 * Statement DTOs. Mirror `org.jacodb.ets.dto` statements
 * (jacodb-ets/src/main/kotlin/org/jacodb/ets/dto/Stmts.kt).
 *
 * Serialized polymorphically with the "_" discriminator.
 * Unknown discriminators fall back to RawStmtDto on the Kotlin side.
 */

import { CallExprDto, ConditionExprDto, ValueDto } from "./values";

/**
 * Closed union of the known statement kinds.
 * Raw fallback statements (RawStmtDto) are intentionally NOT part of this union
 * (index signature breaks narrowing); emission sites cast them explicitly.
 */
export type StmtDto =
    | NopStmtDto
    | AssignStmtDto
    | CallStmtDto
    | ReturnVoidStmtDto
    | ReturnStmtDto
    | ThrowStmtDto
    | IfStmtDto;

/** Fallback for constructs we cannot model. */
export interface RawStmtDto {
    readonly _: string; // any kind not listed below
    [extra: string]: unknown;
}

export interface NopStmtDto {
    readonly _: "NopStmt";
}

export interface AssignStmtDto {
    readonly _: "AssignStmt";
    left: ValueDto; // must convert to Local / FieldRef / ArrayRef on the Kotlin side
    right: ValueDto;
}

export interface CallStmtDto {
    readonly _: "CallStmt";
    expr: CallExprDto;
}

export interface ReturnVoidStmtDto {
    readonly _: "ReturnVoidStmt";
}

export interface ReturnStmtDto {
    readonly _: "ReturnStmt";
    arg: ValueDto;
}

export interface ThrowStmtDto {
    readonly _: "ThrowStmt";
    arg: ValueDto;
}

export interface IfStmtDto {
    readonly _: "IfStmt";
    condition: ConditionExprDto;
}

export const RETURN_VOID_STMT: ReturnVoidStmtDto = { _: "ReturnVoidStmt" };

/** Statements that must terminate a basic block (no stmts may follow them). */
export function isTerminatorStmt(stmt: StmtDto): boolean {
    return (
        stmt._ === "ReturnVoidStmt" ||
        stmt._ === "ReturnStmt" ||
        stmt._ === "ThrowStmt" ||
        stmt._ === "IfStmt"
    );
}
